lte/gateway/c/session_manager/ShardTracker.cpp

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <algorithm>
+#include <map>
+
+#include "ShardTracker.h"
+#define MAX_SHARD_SIZE 100
+
+namespace magma {
+
+ShardTracker::ShardTracker() : number_of_shards_(0) {}
+
+int ShardTracker::add_ue() {
+  int shard_id;
+  // If we have no shards, we add an entry for the shard representing
+  // of 1 representing a single UE at the new shard*/
+  if (shards_.size() == 0) {
+    shard_id = 0;
+    shards_.push_back(1);
+    number_of_shards_++;
+  } else {
+    // We iterate through all existing shards, if all are full
+    // we create a new shard with quantity 1, otherwise we increment
+    // the quantity of UEs in the latest shard
+    for (auto i = 0; i < number_of_shards_; i++) {
+      if (shards_[i] < MAX_SHARD_SIZE) {
+        shards_[i]++;
```

Comment:
This seems like a suboptimal way for sharding right? You will always fill up the buckets with lower indices first. This seems like bin packing to me. If so we should use a different name for the class.
If your objective is to minimize the number of calls to flows in pipelined then I would definitely call it bin packing. If the objective is to truly shard then a vector is the wrong datastructure.

Reply:
Yes, this is just initial implementation. We want to get some semblance of it merged and working and then we'll move onto better allocation and moving UEs around. Iterative approach